File: kreuzberg/packages/php/src/Types/Metadata/BibtexMetadata.php
<?php

declare(strict_types=1);

namespace Kreuzberg\Types\Metadata;

/**
 * BibTeX bibliography metadata.
 *
 * Contains information about BibTeX entries including citation keys,
 * authors, year ranges, and entry type distribution.
 */
readonly class BibtexMetadata
{
    /**
     * @param string[] $citationKeys
     * @param string[] $authors
     * @param array<string, int>|null $entryTypes
     */
    public function __construct(
        public int $entryCount = 0,
        public array $citationKeys = [],
        public array $authors = [],
        public ?YearRange $yearRange = null,
        public ?array $entryTypes = null,
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        $rawEntryCount = $data['entry_count'] ?? 0;
        $entryCount = is_int($rawEntryCount) ? $rawEntryCount : (is_numeric($rawEntryCount) ? (int) $rawEntryCount : 0);

        /** @var string[] $citationKeys */
        $citationKeys = $data['citation_keys'] ?? [];
        if (!is_array($citationKeys)) {
            $citationKeys = [];
        }

        /** @var string[] $authors */
        $authors = $data['authors'] ?? [];
        if (!is_array($authors)) {
            $authors = [];
        }

        $yearRange = null;
        $yearRangeRaw = $data['year_range'] ?? null;
        if (is_array($yearRangeRaw)) {
            /** @var array<string, mixed> $yearRangeData */
            $yearRangeData = $yearRangeRaw;
            $yearRange = YearRange::fromArray($yearRangeData);
        }

        /** @var array<string, int>|null $entryTypes */
        $entryTypes = $data['entry_types'] ?? null;
        if (!is_array($entryTypes)) {
            $entryTypes = null;
        }

        return new self(
            entryCount: $entryCount,
            citationKeys: $citationKeys,
            authors: $authors,
            yearRange: $yearRange,
            entryTypes: $entryTypes,
        );
    }

    /**
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        $result = [
            'entry_count' => $this->entryCount,
        ];

        if ($this->citationKeys !== []) {
            $result['citation_keys'] = $this->citationKeys;
        }

        if ($this->authors !== []) {
            $result['authors'] = $this->authors;
        }

        if ($this->yearRange !== null) {
            $result['year_range'] = $this->yearRange->toArray();
        }

        if ($this->entryTypes !== null) {
            $result['entry_types'] = $this->entryTypes;
        }

        return $result;
    }
}

File: kreuzberg/packages/elixir/lib/kreuzberg/bibtex_metadata.ex
defmodule Kreuzberg.BibtexMetadata do
  @moduledoc """
  BibTeX bibliography metadata.

  Matches the Rust `BibtexMetadata` struct.

  ## Fields

    * `:entry_count` - Number of BibTeX entries
    * `:citation_keys` - List of citation keys
    * `:authors` - List of author names
    * `:year_range` - Year range of entries
    * `:entry_types` - Map of entry type to count
  """

  @type t :: %__MODULE__{
          entry_count: non_neg_integer(),
          citation_keys: list(String.t()),
          authors: list(String.t()),
          year_range: Kreuzberg.YearRange.t() | nil,
          entry_types: map() | nil
        }

  defstruct entry_count: 0,
            citation_keys: [],
            authors: [],
            year_range: nil,
            entry_types: nil

  @spec from_map(map()) :: t()
  def from_map(data) when is_map(data) do
    %__MODULE__{
      entry_count: data["entry_count"] || 0,
      citation_keys: data["citation_keys"] || [],
      authors: data["authors"] || [],
      year_range: normalize_year_range(data["year_range"]),
      entry_types: data["entry_types"]
    }
  end

  @spec to_map(t()) :: map()
  def to_map(%__MODULE__{} = meta) do
    %{
      "entry_count" => meta.entry_count,
      "citation_keys" => meta.citation_keys,
      "authors" => meta.authors,
      "year_range" => serialize_year_range(meta.year_range),
      "entry_types" => meta.entry_types
    }
  end

  defp normalize_year_range(nil), do: nil
  defp normalize_year_range(%Kreuzberg.YearRange{} = yr), do: yr
  defp normalize_year_range(map) when is_map(map), do: Kreuzberg.YearRange.from_map(map)
  defp normalize_year_range(_), do: nil

  defp serialize_year_range(nil), do: nil
  defp serialize_year_range(%Kreuzberg.YearRange{} = yr), do: Kreuzberg.YearRange.to_map(yr)
  defp serialize_year_range(other), do: other
end
